controller/monitor/node_monitor.go

```diff
@@ -0,0 +1,378 @@
+	DiskStatusMap               map[string]*longhorn.DiskStatus
+	DiskInfoMap                 map[string]*DiskInfo
+	OnDiskReplicaDirectoryNames map[string]map[string]string
+}
+
+type GetDiskInfoHandler func(string) (*util.DiskInfo, error)
+type GetDiskConfig func(string) (*util.DiskConfig, error)
+type GenerateDiskConfig func(string) (*util.DiskConfig, error)
+
+type DiskInfo struct {
+	Entry *util.DiskInfo
+	err   error
+}
+
+func NewNodeMonitor(logger logrus.FieldLogger, ds *datastore.DataStore, node *longhorn.Node, syncCallback func(key string)) (*NodeMonitor, error) {
+	ctx, quit := context.WithCancel(context.Background())
+
+	m := &NodeMonitor{
+		baseMonitor: newBaseMonitor(ctx, quit, logger, ds, NodeMonitorSyncPeriod),
+
+		nodeName: node.Name,
+
+		collectedDataLock: sync.RWMutex{},
+		collectedData:     &NodeMonitorCollectedData{},
+
+		syncCallback: syncCallback,
+
+		getDiskInfoHandler: util.GetDiskInfo,
+		getDiskConfig:      util.GetDiskConfig,
+		generateDiskConfig: util.GenerateDiskConfig,
+	}
+
+	go m.Start()
+
+	return m, nil
+}
+
+func (m *NodeMonitor) Start() {
+	wait.PollImmediateUntil(m.syncPeriod, func() (done bool, err error) {
+		if err := m.SyncCollectedData(); err != nil {
+			m.logger.Errorf("failed to sync data because of %v", err)
+		}
+		return false, nil
+	}, m.ctx.Done())
+}
+
+func (m *NodeMonitor) Close() {
+	m.quit()
+}
+
+func (m *NodeMonitor) GetCollectedData() (interface{}, error) {
+	m.collectedDataLock.RLock()
+	defer m.collectedDataLock.RUnlock()
+
+	data := &NodeMonitorCollectedData{}
+	if err := copier.Copy(data, m.collectedData); err != nil {
+		return nil, errors.Wrapf(err, "failed to copy node monitor collected data")
+	}
+
+	return data, nil
+}
+
+func (m *NodeMonitor) SyncCollectedData() error {
+	node, err := m.ds.GetNode(m.nodeName)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to get longhorn node %v", m.nodeName)
+		return err
+	}
+
+	pruneDiskStatus(node)
+
+	diskStatusMap, diskInfoMap := m.syncDiskStatus(node)
+	onDiksReplicaDirectoryNames := m.getOnDiskReplicaDirectoryNames(node)
+
+	if isDiskStatusChanged(m.collectedData.DiskStatusMap, diskStatusMap) ||
+		!reflect.DeepEqual(m.collectedData.OnDiskReplicaDirectoryNames, onDiksReplicaDirectoryNames) {
+		m.updateCollectedData(diskStatusMap, diskInfoMap, onDiksReplicaDirectoryNames)
+		key := node.Namespace + "/" + m.nodeName
+		m.syncCallback(key)
+	}
+
+	return nil
+}
+
+func (m *NodeMonitor) updateCollectedData(diskStatusMap map[string]*longhorn.DiskStatus, diskInfoMap map[string]*DiskInfo, onDiskReplicaDirectoryNames map[string]map[string]string) {
+	m.collectedDataLock.Lock()
+	defer m.collectedDataLock.Unlock()
+
+	m.collectedData.DiskStatusMap = diskStatusMap
+	m.collectedData.DiskInfoMap = diskInfoMap
+	m.collectedData.OnDiskReplicaDirectoryNames = onDiskReplicaDirectoryNames
+}
+
+func (m *NodeMonitor) syncDiskStatus(node *longhorn.Node) (diskStatusMap map[string]*longhorn.DiskStatus, diskInfoMap map[string]*DiskInfo) {
```

Comment:
Then this function can be like:
1. Load the disk info
2. Get or (blindly) generate the disk UUID

Here, I prefer to ask the node controller itself to handle the duplicate fsid issue. Since this monitor actually does not know which disk should be marked as NotReady or not when multiple paths hold the same fsid. 
For example, there is only one disk on the node at the beginning and everything works fine. Later on, users add some sub-directories of the current disk path as the new disks for the node (and some of the subdir contain DiskUUID). Then Longhorn should mark all disks **except for 1st one** as NotReady. The monitor does not know the previous disk status info (which disk is already considered Ready) hence it cannot handle this case.

Reply:
Okay. The fsid handling will be moved to node controller.